common/configurable_model.py

```diff
@@ -0,0 +1,38 @@
+from bit_vector import BitVector
+from common.model import Model
+
+
+def ConfigurableModel(config_data_width, config_addr_width):
+    """
+    This function returns a class (parameterized by @config_data_width and
+    @config_addr_width) which can be used as the base class for any functional
+    model implementing a configurable module. For example if we have
+    configurable module Foo:
+
+        class Foo(ConfigurableModel(32, 32)):
+            ...
+
+    In our test code we can then do:
+
+        foo = Foo(...)
+        foo.configure(addr, data)
+        assert foo.read_config(addr) == data
+
+    Or internally in Foo, do:
+
+        conf = self._config[addr]
+
+    """
+    class _ConfigurableModel(Model):
+        def __init__(self):
+            self._config = {}
+
+        def configure(self, addr: BitVector, data: BitVector):
+            assert data.num_bits == config_data_width
+            assert addr.num_bits == config_addr_width
+            self._config[addr.unsigned_value] = data
+
+        def read_config(self, addr):
+            return self._config[addr.unsigned_value]
```

Comment:
> From @leonardt  (#29):
> 
> What about using a `@property` to implement a getter/setter interface like
> 
> ```
> foo.config[addr] = data
> ```
> 
> and
> 
> ```
> conf = foo.config[addr]
> ```
> 
> Here's some docs on property: https://docs.python.org/3.6/library/functions.html#property

Reply:
Generally I like the `@property` paradigm. But then we lose control over the internals of the setter right? E.g. we can't enforce that `addr` and `data` have the proper widths?

Reply:
Oh, now that I think about it, to achieve the interface I described, you could make `config` a dictionary, and override `__setitem__` and `__getitem__`. Here's an example from `pycoreir` https://github.com/leonardt/pycoreir/blob/master/coreir/context.py#L24-L42

Reply:
That examples allows you to write
```
context.named_types[("coreir", "clk")]
```
which actually calls into the coreir API to get the namedtype

Reply:
😄 I ended up doing something similar.